rust/lance-index/src/vector/pq.rs

```diff
@@ -340,23 +340,7 @@ impl<T: ArrowFloatType + Cosine + Dot + L2 + 'static> ProductQuantizer for Produ
 
         let fsl = if self.metric_type == MetricType::Cosine {
             // Normalize cosine vectors to unit length.
-            let values = fsl
-                .values()
-                .as_any()
-                .downcast_ref::<T::ArrayType>()
-                .ok_or(Error::Index {
-                    message: format!(
-                        "Expect to be a float vector array, got: {:?}",
-                        fsl.value_type()
-                    ),
-                    location: location!(),
-                })?
-                .as_slice()
-                .chunks(self.dimension)
-                .flat_map(normalize)
-                .collect::<Vec<_>>();
-            let data = T::ArrayType::from(values);
-            FixedSizeListArray::try_new_from_values(data, self.dimension as i32)?
+            normalize_fsl(&fsl)?
```

Comment:
Should we remove cosine type PQ entirely? And let `IVF` handle normalization and resid?

Reply:
^^ can we add a `panic!` that make sure we don't create a cosine PQ?